src/Microsoft.OData.Client/Serialization/HttpWebRequestMessage.cs

```diff
@@ -307,158 +379,101 @@ public override IAsyncResult BeginGetResponse(AsyncCallback callback, object sta
             {
-                //throw new Exception("pppp test" +this.httpRequest.Address+" "+webException.Message);
-                throw ConvertToDataServiceWebException(webException);
-            }
+                var send = CreateSendTask();
+                send.Wait();
+                _httpResponseMessage = send.Result;
+
+                return ConvertHttpClientResponse(send.Result);
+            });
         }
-#endif
 
-        /// <summary>
-        /// Sets the Content length of the Http web request
-        /// </summary>
-        /// <param name="httpWebRequest">Http web request to set the content length on</param>
-        /// <param name="contentLength">Length to set</param>
-        internal static void SetHttpWebRequestContentLength(HttpWebRequest httpWebRequest, long contentLength)
+        private Task<HttpResponseMessage> CreateSendTask()
         {
-#if PORTABLELIB
-            // TODO: May need to add a special case for when the content length is zero
-            bool headerSet = false;
-            if (ContentLengthMethodSetter != null && SettingContentLengthPropertyValidOnCurrentPlatform)
+            // Only set the message content if the stream has been written to, otherwise
+            // HttpClient will complain if it's a GET request.
+            var messageContent = _messageStream.ToArray();
+            if (messageContent.Length > 0)
             {
-                try
-                {
-                    ContentLengthMethodSetter.Invoke(httpWebRequest, new object[] { contentLength });
-                    headerSet = true;
-                }
-                catch (InvalidOperationException)
-                {
-                    SettingContentLengthPropertyValidOnCurrentPlatform = false;
-                }
-            }
+                _requestMessage.Content = new ByteArrayContent(messageContent);
 
-            if (!headerSet && SettingContentLengthOnHeadersCollectionValidOnCurrentPlatform)
-            {
-                // Attempting to set the header but may just not work
-                try
-                {
-                    // Need this code to set the user agent on silverlight when the property doesn't exist
-                    httpWebRequest.Headers[XmlConstants.HttpContentLength] = XmlConvert.ToString(contentLength);
-                }
-                catch (ArgumentException)
+                // Apply cached "Content" header values
+                foreach (var contentHeader in _contentHeaderValueCache)
                 {
-                    SettingContentLengthOnHeadersCollectionValidOnCurrentPlatform = false;
+                    _requestMessage.Content.Headers.Add(contentHeader.Key, contentHeader.Value);
                 }
             }
-#else
-            httpWebRequest.ContentLength = contentLength;
-#endif
+
+            _requestMessage.Method = new HttpMethod(_effectiveHttpMethod);
+
+            var send = _client.SendAsync(_requestMessage);

```

Comment:
Other than in test projects, the practice is avoid using `var`. Either do:
```csharp
return _client.SendAsync(_requestMessage);
```
or
```csharp
Task<HttpResponseMessage> sendTask = _client.SendAsync(_requestMessage);
return sendTask;
```